lib/membrane_ffmpeg_video_filter/text_overlay.ex

```diff
@@ -0,0 +1,157 @@
+
+  The element expects each frame to be received in a separate buffer.
+  Additionally, the element has to receive proper caps with picture format and dimensions.
+  """
+  use Membrane.Filter
+  alias __MODULE__.Native
+  alias Membrane.Buffer
+  alias Membrane.Caps.Video.Raw
+
+  def_options text: [
+                type: :binary,
+                description: "Text to be displayed on video"
+              ],
+              fontsize: [
+                type: :int,
+                description:
+                  "Size of the displayed font. If not set, ffmpeg's default value (16) is applied",
+                default: nil
+              ],
+              fontcolor: [
+                type: :binary,
+                description:
+                  "Choose font color according to the ffmpeg color syntax (https://ffmpeg.org/ffmpeg-utils.html#color-syntax).
+                  If not set, ffmpeg's default value (black) is applied",
+                default: nil
+              ],
+              fontfile: [
+                type: :binary,
+                description:
+                  "Path to the file with the desired font. If not set, default font fallback from fontconfig is used",
+                default: nil
+              ],
+              box?: [
+                type: :boolean,
+                description: "Set to true if a box is to be displayed behind the text",
+                default: false
+              ],
+              boxcolor: [
+                type: :binary,
+                description: "If the box? is set to true, display a box in the given color",
+                default: nil
+              ],
+              border?: [
+                type: :boolean,
+                description: "Set to true to display a gray border around letters",
+                default: false
+              ],
+              x: [
+                type: :atom,
+                spec: :left | :right | :center,
+                description: "Horizontal position of the displayed text",
+                default: :left
+              ],
+              y: [
+                type: :atom,
+                spec: :top | :bottom | :center,
+                description: "Vertical position of the displayed text",
+                default: :bottom
+              ]
+
+  def_input_pad :input,
+    demand_unit: :buffers,
+    caps: {Raw, aligned: true}
+
+  def_output_pad :output,
+    caps: {Raw, aligned: true}
+
+  @impl true
+  def handle_init(options) do
+    state =
+      options
+      |> Map.from_struct()
+      |> Map.put(:native_state, nil)
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_demand(:output, _size, :buffers, _ctx, %{native_state: nil} = state) do
+    # Wait until we have native state (after receiving caps)
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_demand(:output, size, :buffers, _ctx, state) do
+    {{:ok, demand: {:input, size}}, state}
+  end
+
+  @impl true
+  def handle_process(
```

Comment:
Add a clause raising on buffer arriving before caps (i.e. `native_state` being `nil`)

Reply:
Currently the element doesn't send demand until caps arrived and state is initialized, as is in our h264 encoder (https://github.com/membraneframework/membrane_h264_ffmpeg_plugin/blob/master/lib/membrane_h264_ffmpeg/encoder.ex). Doesn't that prevent buffer from arriving before caps? Or should I still add a clause in handle_process?

`@impl true
  def handle_demand(:output, _size, :buffers, _ctx, %{native_state: nil} = state) do
    # Wait until we have native state (after receiving caps)
    {:ok, state}
  end
`

Reply:
You're right, that's enough. We're going to enforce that at framework level anyway (https://github.com/membraneframework/membrane_core/issues/341)